File: csv_detective/formats/geojson.py
import json

proportion = 1
description = "JSON object in the [GeoJSON](https://fr.wikipedia.org/wiki/GeoJSON) format"
tags = ["geo"]
python_type = "json"
labels = {
    "json geojson": 1,
    "json": 1,
    "geojson": 1,
    "geo shape": 1,
    "geom": 0.75,
    "geometry": 1,
    "geoshape": 1,
}


def _check_dict(d: dict) -> bool:
    if "type" in d and "coordinates" in d:
        return True
    if "geometry" in d and "coordinates" in d["geometry"]:
        return True
    return False


def _is(val) -> bool:
    if isinstance(val, dict):
        return _check_dict(val)
    try:
        j = json.loads(val)
        if isinstance(j, dict):
            return _check_dict(j)
    except Exception:
        pass
    return False


_test_values = {
    True: [
        '{"coordinates": [45.783753, 3.049342], "type": "63870"}',
        '{"geometry": {"coordinates": [45.783753, 3.049342]}}',
        {"geometry": {"coordinates": [45.783753, 3.049342]}},
        {"coordinates": [45.783753, 3.049342], "type": "63870"},
    ],
    False: ['{"pomme": "fruit", "reponse": 42}'],
}

File: rust/src/formats/geojson.rs
use super::Detector;

pub struct GeoJsonFormat;

impl GeoJsonFormat {
    pub fn detect(&self, val: &str) -> Option<()> {
        let trimmed = val.trim();
        if !trimmed.starts_with('{') {
            return None;
        }
        let obj: serde_json::Value = serde_json::from_str(trimmed).ok()?;
        let map = obj.as_object()?;
        // either has type+coordinates, or geometry.coordinates
        if map.contains_key("type") && map.contains_key("coordinates") {
            return Some(());
        }
        if let Some(geom) = map.get("geometry") {
            if let Some(g) = geom.as_object() {
                if g.contains_key("coordinates") {
                    return Some(());
                }
            }
        }
        None
    }
}

impl Detector for GeoJsonFormat {
    fn name(&self) -> &'static str { "geojson" }
    fn python_type(&self) -> &'static str { "json" }
    fn proportion(&self) -> f64 { 1.0 }
    fn tags(&self) -> &'static [&'static str] { &["geo"] }
    fn labels(&self) -> &'static [(&'static str, f64)] {
        &[
            ("json geojson", 1.0), ("json", 1.0), ("geojson", 1.0),
            ("geo shape", 1.0), ("geom", 0.75), ("geometry", 1.0), ("geoshape", 1.0),
        ]
    }
    fn test(&self, val: &str) -> bool { self.detect(val).is_some() }
}
